File: server/src/main/java/soze/multilife/simulation/Grid.java
package soze.multilife.simulation;

import soze.multilife.simulation.rule.Rule;

import java.awt.*;
import java.util.*;
import java.util.List;
import java.util.stream.Collectors;

/**
 * A grid of cells. This class manages the cells, lets the simulation
 * set cell states.
 */
public class Grid {

  /**
   * This map contains all cells.
   */
  private final Map<Point, Cell> cells = new HashMap<>();
  /**
   * This map contains current active cells. During a grid update,
   * only cells contained within this collection are checked.
   */
  private final Map<Point, Cell> activeCells = new HashMap<>();
  /**
   * This map contains cells which will become active cells
   * in the next iteration.
   */
  private final Map<Point, Cell> nextCells = new HashMap<>();

  /**
   * Contains rules for players.
   */
  private final Map<Long, Rule> rules = new HashMap<>();
  private final int width;
  private final int height;

  Grid(int width, int height) {
	this.width = width;
	this.height = height;
	init();
  }

  /**
   * Populates the grid of all cells.
   */
  private void init() {
	for (int i = 0; i < width; i++) {
	  for (int j = 0; j < height; j++) {
		cells.put(new Point(i, j), new Cell(i, j));
	  }
	}
  }

  /**
   * Adds a rule for a given player.
   *
   * @param ownerId id of the player this rule should apply to
   * @param rule    rule to apply to cells. Cannot be null.
   */
  void addRule(long ownerId, Rule rule) {
	rules.put(ownerId, Objects.requireNonNull(rule));
  }

  /**
   * @return all cells in this grid
   */
  Collection<Cell> getAllCells() {
	return cells.values();
  }

  /**
   * Changes state of a cell at location x, y.
   *
   * @param x       x location of the cell
   * @param y       y location of the cell
   * @param state   state of the cell
   * @param ownerId owner of the cell
   */
  void changeState(int x, int y, boolean state, long ownerId) {
	Point p = getPoint(x, y);
	changeState(p, state, ownerId);
  }


  /**
   * Adds all given cells to the next cells.
   * @param cells
   */
  void click(Collection<Cell> cells) {
	for(Cell cell: cells) {
	  nextCells.put(new Point(cell.getX(), cell.getY()), cell);
	}
  }

  /**
   * Finds all cells that are clickable by this player.
   * @param indices array of indices of cells
   * @param ownerId id of the player who clicked
   * @return a list of all clickable (by this player) \s
   */
  List<Cell> findClickableCells(int[] indices, long ownerId) {
    List<Cell> clickableCells = new ArrayList<>();
	for(int i: indices) {
	  Point p = getPoint(i);
	  Cell cell = cells.get(p);
	  if(!cell.isAlive()) {
		cell = new Cell(p.x, p.y);
		cell.setIsAlive(true);
		cell.setOwnerId(ownerId);
		clickableCells.add(cell);
	  }
	}
	return clickableCells;
  }

  /**
   * Changes the state of a cell at a given point.
   *
   * @param p point containing coordinates
   * @param state alive/dead
   * @param ownerId id of the owner
   */
  private void changeState(Point p, boolean state, long ownerId) {
	Cell cell = nextCells.get(p);
	if (cell == null) {
	  cell = new Cell(p.x, p.y);
	  nextCells.put(p, cell);
	}
	cell.setIsAlive(state);
	cell.setOwnerId(ownerId);
  }

  /**
   * Adds this cell to the map of active cells.
   * This method also adds neighbours of this cell to active cells.
   * If cells already exist, cells are not replaced.
   *
   * @param cell
   */
  private void addToActive(Cell cell) {
	int x = cell.getX();
	int y = cell.getY();
	for (int i = -1; i < 2; i++) {
	  for (int j = -1; j < 2; j++) {
		Point p = getPoint(i + x, j + y);
		activeCells.putIfAbsent(p, getCell(p.x, p.y));
	  }
	}
  }

  /**
   * Returns a cell at this position. If x or y is out of bounds,
   * this method will wrap around the grid.
   *
   * @param x
   * @param y
   * @return cell at given coordinates (from all cells)
   */
  private Cell getCell(int x, int y) {
	return cells.get(getPoint(x, y));
  }

  /**
   * Updates the cells and transfers the cells from nextCells to active cells
   * and updates the underlying map.
   */
  void updateGrid() {
    update();
    transferCells();
  }

  /**
   * Goes through active cells and populates next iteration
   * active cells.
   */
  private void update() {
	for (Cell cell : activeCells.values()) {
	  int x = cell.getX();
	  int y = cell.getY();
	  List<Cell> aliveNeighbours = getAliveNeighbourCells(x, y);
	  long ownerId = cell.getOwnerId();
	  int state = rules.get(ownerId).apply(aliveNeighbours.size(), cell.isAlive());
	  if (state != 0) {
		long strongestOwnerId = getStrongestOwnerId(aliveNeighbours);
		changeState(x, y, state > 0, strongestOwnerId == -1 ? cell.getOwnerId() : strongestOwnerId);
	  }
	}
  }

  /**
   * Returns all alive cells around a cell at location x, y.
   *
   * @param x
   * @param y
   * @return
   */
  private List<Cell> getAliveNeighbourCells(int x, int y) {
	List<Cell> aliveNeighbourCells = new ArrayList<>();
	for (int i = -1; i < 2; i++) {
	  for (int j = -1; j < 2; j++) {
		if (i == 0 && j == 0) continue;
		Cell cell = getCell(x + i, y + j);
		if (cell.isAlive()) aliveNeighbourCells.add(cell);
	  }
	}
	return aliveNeighbourCells;
  }

  /**
   * Finds the most frequently (mode) occuring ownerId among given cells.
   * If there are no cells, returns -1.
   * @param cells
   * @return
   */
  private long getStrongestOwnerId(List<Cell> cells) {
    if(cells.isEmpty()) {
      return -1;
	}
	List<Long> ownerIds = cells.stream()
	  .map(Cell::getOwnerId)
	  .collect(Collectors.toList());
	return mode(ownerIds);
  }

  private long mode(List<Long> list) {
	long maxValue = 0, maxCount = 0;

	for (int i = 0; i < list.size(); ++i) {
	  int count = 0;
	  for (int j = 0; j < list.size(); ++j) {
		if (list.get(j).equals(list.get(i))) ++count;
	  }
	  if (count > maxCount) {
		maxCount = count;
		maxValue = list.get(i);
	  }
	}

	return maxValue;
  }

  /**
   * Switches active cells and next iteration active cells.
   * After the switch, next iteration cells are cleared.
   */
  void transferCells() {
	activeCells.clear();
	for (Cell nextCell : nextCells.values()) {
	  Cell cell = getCell(nextCell.getX(), nextCell.getY());
	  cell.setIsAlive(nextCell.isAlive());
	  cell.setOwnerId(nextCell.getOwnerId());
	  addToActive(cell);
	}
	nextCells.clear();
  }

  /**
   * Creates a point object with a given x, y values.
   * If the point lies out of bounds then x, y values
   * are wrapped around.
   *
   * @return
   */
  private Point getPoint(int x, int y) {
	int index = getIndex(x, y);
	int wrappedX = index % width;
	int wrappedY = index / width;
	return new Point(wrappedX, wrappedY);
  }

  private int getIndex(int x, int y) {
	int index = x + (y * width); // find index
	return wrapIndex(index); // wrap it
  }

  private int wrapIndex(int index) {
	int maxSize = width * height;
	if (index < 0) return index + (maxSize); // wrap around if neccesary
	if (index >= maxSize) return index % (maxSize);
	return index; // return index if not neccesary to wrap
  }

  /**
   * Returns x, y coordinates of a cell at a given index,
   * as if the cells were contained in an 1D array.
   * The coordinates are returned in a Point object.
   *
   * @param index
   * @return
   */
  private Point getPoint(int index) {
    index = wrapIndex(index);
	return new Point(index % width, index / height);
  }


}

File: client/grid.js
class Grid {

    constructor(width, height, playerData) {
        this.width = width;
        this.height = height;
        this.cells = {};
        this.activeCells = {};
        this.nextCells = {};
        this.playerData = playerData;
        this.cellSize = 10;
        this.rules = new Rules();
        this.init();
    }

    /**
     * Creates all cells.
     */
    init() {
        for(let i = 0; i < width; i++) {
            for(let j = 0; j < height; j++) {
                this.cells["x:" + i + "y:" + j] = new Cell(i, j, false, 0, this.cellSize, this.getColor(0), Cell.ellipseRenderFunction);
            }
        }
    }

    /**
        Sets the state of a cell at a given position.
        The cell is in nextCells object.
    */
    setCellState(position, alive, ownerId) {
        let cell = this.nextCells[this.getPositionKey(position)];
        if(!cell) {
            cell = new Cell(position.x, position.y, alive, ownerId, this.cellSize, this.getColor(ownerId), Cell.ellipseRenderFunction);
            this.nextCells[this.getPositionKey(position)] = cell;
        }
    }

    /**
        Transforms the position of a cell into the key used in maps
        containing the cells.
    */
    getPositionKey(position) {
        let x = position.x;
        let y = position.y;
        if(x < 0) {
            x = this.width - 1;
        }
        if(x === this.width) {
            x = 0;
        }
        if(y < 0) {
            y = this.height - 1;
        }
        if(y === this.height) {
            y = 0;
        }
        return "x:" + x + "y:" + y;
    }

    getColor(ownerId) {
        return this.playerData.colors[ownerId] || "#000000";
    }

    updateCells() {
        for(let pos in this.activeCells) {
            if (this.activeCells.hasOwnProperty(pos)) {
                let cell = this.activeCells[pos];
                let x = cell.x;
                let y = cell.y;
                let aliveNeighbours = this.getAliveNeighbourCells(x, y);
                let ownerId = cell.getOwnerId();
                let state = this.rules.getRule(this.playerData.rules[ownerId])(aliveNeighbours.length, cell.isAlive());
                if (state != 0) {
                    let strongestOwnerId = this.getStrongestOwnerId(aliveNeighbours);
                    this.setCellState({x:x, y:y}, state > 0, strongestOwnerId == -1 ? cell.getOwnerId() : strongestOwnerId);
                }
            }
        }
    }

    /**
        Returns an array of alive cells neighbouring a cell at x, y.
    */
    getAliveNeighbourCells(x, y) {
       let cells = [];
       for(let i = -1; i < 2; i++) {
           for(let j = -1; j < 2; j++) {
               if(i === 0 && j === 0) continue;
               let positionKey = this.getPositionKey({x: i + x, y: j + y});
               let cell = this.cells[positionKey];
               if(cell.isAlive()) cells.push(cell);
           }
       }
       return cells;
    }

    getStrongestOwnerId(cells) {
        if(cells.length === 0) {
            return -1;
        }

        let ownerIds = cells.map(function(cell, index) {
           return cell.getOwnerId();
        });

        return this.mode(ownerIds);
    }

    mode(ownerIds) {
        let maxValue = 0, maxCount = 0;

        for (let i = 0; i < ownerIds.length; ++i) {
            let count = 0;
            for (let j = 0; j < ownerIds.length; ++j) {
                if (ownerIds[j] === ownerIds[i]) ++count;
            }
            if (count > maxCount) {
                maxCount = count;
                maxValue = ownerIds[i];
            }
        }
        return maxValue;
    }

    transferCells() {
        this.activeCells = {};
        for(let pos in this.nextCells) {
            if(this.nextCells.hasOwnProperty(pos)) {
                let oldCell = this.nextCells[pos];
                let c = this.cells[pos];
                c.setAlive(oldCell.isAlive());
                c.setOwnerId(oldCell.getOwnerId());
                c.setColor(this.getColor(oldCell.getOwnerId()));
                this.addToActive(c);
            }
        }
        this.nextCells = {};
    }

    addToActive(cell) {
        let x = cell.x;
        let y = cell.y;
        for(let i = -1; i < 2; i++) {
            for(let j = -1; j < 2; j++) {
                let position = this.getPositionKey({x: i + x, y: j + y});
                let cell = this.activeCells[position];
                if(!cell) {
                    this.activeCells[position] = this.cells[position];
                }
            }
        }

    }

    render(viewport) {
        for(let pos in this.cells) {
            if(this.cells.hasOwnProperty(pos)) {
				this.cells[pos].update();
                this.cells[pos].render(viewport);
            }
        }
    }

    setPlayerData(playerData) {
        this.removeDisconnectedPlayers(playerData);
        Object.assign(this.playerData, playerData);
    }

    removeDisconnectedPlayers(newPlayerData) {
        let oldData = this.playerData;
        let oldPlayerIds = this.getPlayerIds(oldData.rules);
        let newPlayerIds = this.getPlayerIds(newPlayerData.rules);
        let disconnectedPlayers = this.findDisconnectedPlayers(oldPlayerIds, newPlayerIds);
        for(let i = 0; i < disconnectedPlayers.length; i++) {
            let id = disconnectedPlayers[i];
            for(let pos in this.cells) {
                if(this.cells.hasOwnProperty(pos)) {
                    let cell = this.cells[pos];
                    if(cell.ownerId == id) {
                        cell.setAlive(false);
                        cell.setOwnerId(0);
                    }
                }
            }
        }
    }

    getPlayerIds(playerData) {
        let arr = [];
        for(let id in playerData) {
            if(playerData.hasOwnProperty(id)) {
                arr.push(id);
            }
        }
        return arr;
    }

    findDisconnectedPlayers(oldPlayerIds, newPlayerIds) {
        let disconnectedPlayers = [];
        for(let i = 0; i < oldPlayerIds.length; i++) {
            let oldPlayerId = oldPlayerIds[i];
            let index = newPlayerIds.findIndex((i) => i === oldPlayerId);
            if(index === -1) {
                disconnectedPlayers.push(oldPlayerId);
            }
        }
        return disconnectedPlayers;
    }

}
